### PR TITLE
Add support for maintenance windows of auto synced applications

### PR DESCRIPTION
**Pull request by adamjohnson01**
This is related to #1995, it will prevent auto sync during specified time windows but will still allow a manual sync.

Windows are configured using a cron format with a duration.

```
syncPolicy:
  automated:
    prune: true
    maintenanceWindows:

    # Tuesdays at 2am-4am
    - schedule: "0 2 * * 2" 
      duration: 2h

    # Any time on the first of every even-numbered month
    - schedule: "0 0 1 2 *"
      duration: 24h
    - schedule: "0 0 1 4 *"
      duration: 24h
    - schedule: "0 0 1 6 *"
      duration: 24h
    - schedule: "0 0 1 8 *"
      duration: 24h
    - schedule: "0 0 1 10 *"
      duration: 24h
    - schedule: "0 0 1 12 *"
      duration: 24h
```

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -437,6 +437,16 @@ type SyncPolicyAutomated struct {
...
 	// SelfHeal enables auto-syncing if  (default: false)
 	SelfHeal bool `json:"selfHeal,omitempty" protobuf:"bytes,2,opt,name=selfHeal"`
+	// MaintenanceWindows disables auto-syncing during the configured time windows  (default: false)
+	MaintenanceWindows []*MaintenanceWindow `json:"maintenanceWindows,omitempty" protobuf:"bytes,3,opt,name=maintenanceWindows"`
+}
+
+// MaintenanceWindow controls when automated syncs should be disabled
+type MaintenanceWindow struct {
+	// Schedule is the time the window will begin
+	Schedule *string `json:"schedule,omitempty" protobuf:"bytes,1,opt,name=schedule"`
```
Comment (alexec) [inline pkg/apis/application/v1alpha1/types.go]: do you want to state this is cron format?

File: ui/src/app/applications/components/application-summary/application-summary.tsx [hunk 2]
```diff
@@ -224,16 +227,26 @@ export const ApplicationSummary = (props: {
...
                                     ) || (
                                         <button className='argo-button argo-button--base' onClick={() => setAutoSync(
                                             ctx, 'Enable Self Heal?', 'Are you sure you want to enable automated self healing?',
-                                            app.spec.syncPolicy.automated.prune, true)
+                                            app.spec.syncPolicy.automated.prune, true, app.spec.syncPolicy.automated.maintenanceWindows)
                                         }>Enable</button>
                                     )}
                                 </div>
                             </div>
+                            {app.spec.syncPolicy.automated.maintenanceWindows && (
```
Comment (alexec) [inline ui/src/app/applications/components/application-summary/application-summary.tsx]: any chance of a screenshot attached to the PR?

File: pkg/apis/application/v1alpha1/types.go [hunk 3]
```diff
@@ -437,6 +437,16 @@ type SyncPolicyAutomated struct {
 	Prune bool `json:"prune,omitempty" protobuf:"bytes,1,opt,name=prune"`
 	// SelfHeal enables auto-syncing if  (default: false)
 	SelfHeal bool `json:"selfHeal,omitempty" protobuf:"bytes,2,opt,name=selfHeal"`
+	// MaintenanceWindows disables auto-syncing during the configured time windows  (default: false)
+	MaintenanceWindows []*MaintenanceWindow `json:"maintenanceWindows,omitempty" protobuf:"bytes,3,opt,name=maintenanceWindows"`
```
Comment (alexec) [inline pkg/apis/application/v1alpha1/types.go]: I love to create `type MaintenanceWindows = []*MaintenanceWindow` as you can then create func for it.

**Review by alexec — CHANGES_REQUESTED** [review]
Wow! Fantastic!! From zero to Argo CD expert in one PR. I've a few comments.

* I think you can encapsulate much of the parsing login into receiver func which will be easier to test and easier to reuse. There are a few examples in `types.go` that allow you to add extra logic too, e.g. `func (in *ApplicationSourceHelm) AddParameter(p HelmParameter)` or `type ResourceResults []*ResourceResult`. These suggestions are optional, but I think they would make the code easier to work with in the future.
* Do you want to add a doc file for this, e.g. `docs/user-guide/maintenance-windows.md`, which can contain a screenshot and example explaining the feature we can then use to promote it in the community.
* I've triggered a CI re-build, but e2e tests that failed may be related to you changes (rather than it just being the usual flaky build). We'll see shortly...

If you want to chat on Slack, come find me in https://argoproj.slack.com/messages/CEWCP8TV5

File: ui/src/app/applications/components/application-summary/application-summary.tsx (hunk 2 above)
Comment (adamjohnson01) [reply ui/src/app/applications/components/application-summary/application-summary.tsx]: Screenshot 2019-08-22 at 08 47 25

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (adamjohnson01) [reply pkg/apis/application/v1alpha1/types.go]: Sure, will do!

**Comment by adamjohnson01** [comment]
I am closing this in favour of #2191

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -257,15 +257,22 @@ func printAppSummaryTable(app *argoappv1.Application, appURL string) {
 	fmt.Printf(printOpFmtStr, "Path:", app.Spec.Source.Path)
 	printAppSourceDetails(&app.Spec.Source)
 	var syncPolicy string
+	var maintenanceWindows string
 	if app.Spec.SyncPolicy != nil && app.Spec.SyncPolicy.Automated != nil {
 		syncPolicy = "Automated"
 		if app.Spec.SyncPolicy.Automated.Prune {
 			syncPolicy += " (Prune)"
 		}
+		if app.Spec.SyncPolicy.Automated.MaintenanceWindows != nil {
```
[the remark this answers is not in the document]
Comment (adamjohnson01) [reply cmd/argocd/commands/app.go]: I have done this in the new pull request.

File: cmd/argocd/commands/app.go [hunk 5]
```diff
@@ -486,6 +505,20 @@ func setKustomizeImages(src *argoappv1.ApplicationSource, images []string) {
 	}
 }
 
+func setMaintenanceWindows(app *argoappv1.Application, windows string) {
+	var maintenanceWindows []*argoappv1.MaintenanceWindow
+	wn := strings.Split(windows, ",")
+	for _, w := range wn {
+		cfg := strings.Split(w, ":")
+		window := argoappv1.MaintenanceWindow{
```
[the remark this answers is not in the document]
Comment (adamjohnson01) [reply cmd/argocd/commands/app.go]: I have done this in the new pull request.

File: cmd/argocd/commands/app.go [hunk 6]
```diff
@@ -486,6 +505,20 @@ func setKustomizeImages(src *argoappv1.ApplicationSource, images []string) {
...
+			Schedule: &cfg[0],
+			Duration: &cfg[1],
+		}
+		maintenanceWindows = append(maintenanceWindows, &window)
```
[the remark this answers is not in the document]
Comment (adamjohnson01) [reply cmd/argocd/commands/app.go]: I have done this in the new pull request.

File: cmd/argocd/commands/app.go [hunk 7]
```diff
@@ -1052,6 +1089,26 @@ func formatSyncPolicy(app argoappv1.Application) string {
 	return policy
 }
 
+func formatMaintenanceWindows(app argoappv1.Application) string {
```
[the remark this answers is not in the document]
Comment (adamjohnson01) [reply cmd/argocd/commands/app.go]: I have done this in the new pull request.

File: controller/appcontroller.go [hunk 8]
```diff
@@ -1117,3 +1129,24 @@ func toggledAutomatedSync(old *appv1.Application, new *appv1.Application) bool {
 	// nothing changed
 	return false
 }
+
+func maintenanceWindowActive(app *appv1.Application) (bool, error) {
```
[the remark this answers is not in the document]
Comment (adamjohnson01) [reply controller/appcontroller.go]: I have done this in the new pull request.

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (adamjohnson01) [reply pkg/apis/application/v1alpha1/types.go]: I have done this in the new pull request.

File: pkg/apis/application/v1alpha1/types.go (hunk 3 above)
Comment (adamjohnson01) [reply pkg/apis/application/v1alpha1/types.go]: I have done this in the new pull request.

**Pull request closed without merging**